infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision.env;
+
+import static java.lang.String.format;
+import static java.util.Comparator.naturalOrder;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.fabric8.kubernetes.api.model.EnvVar;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+/**
+ * If there are any references to other environment variables in the value of some environment
+ * variable, the "dependent" environment variable needs to be defined "later" than the env var it
+ * depends on for Kubernetes variable expansion to work.
+ *
+ * <p>This class makes sure that the environment variables are sorted.
+ */
+final class EnvironmentVariablesSorter implements Comparator<EnvVar> {
+
+  private static final Pattern REFERENCE_PATTERN = Pattern.compile("\\$\\(\\w+\\)");
+
+  private final Map<String, SortedSet<String>> referenceMap;
+  private final Map<String, SortedSet<String>> recursiveReferenceMap;
+
+  private EnvironmentVariablesSorter(Map<String, EnvVar> envVars) {
+    this.referenceMap = new HashMap<>(envVars.size());
+    for (Map.Entry<String, EnvVar> e : envVars.entrySet()) {
+      referenceMap.put(e.getKey(), findReferences(e.getValue()));
```

Comment:
So this class is a comparator, but it needs some initial set up. For the cycle detection to work, it needs to "see" the full list of environment variables to build the reference map eagerly so that it can be effectively used during the comparsions.

The static method is merely a helper to do the two things together - initialize the comparator and let it sort the environment variables.

At the same time I made the constructor private because it doesn't make sense to keep the instance around - the comparator is only valid for a concrete list of environment variables and cannot be reused for any other (because of the eager init of the internal state based on the list of env vars).

I could theoretically split this into 2 classes - the comparator and some new utitility class only exposing the single static method that would be the same as `sort` but I thought that would be an unnecessary overkill..

Reply:
Since the class is not used as comparator elsewhere I do not think it can be treated as a comparator.  The same could be achieved using private inner class without SRP violation